### ISSUE TITLE
script that pulls a group of data packages into a sqlite database

### ISSUE THREAD

**Issue by cmgosnell** [catalyst-cooperative/pudl#364]
In order to do this, I believe first we'll need to squish the multi-package bundle back into one package. This will require having more of a set list of possible data packages and potentially removing duplicate tables/data sets when necessary (i.e. eia tables in an epa cems data pacakge). It will also require pulling the metadata together into one set of metadata.. which will also require not including duplicative tabular resource descriptors.

Once the bundle of data packages are all in one place, writing this script will be easy peasy.

**Comment by cmgosnell** [comment]
hey @roll! I'm getting a cast error when running saving to sqlite when one of the tables has been partitioned.
```
pkg = Package(str(pathlib.Path(pudl_settings['datapackage_dir'],
                                   pkg_bundle_dir_name,
                                   pkg_name, 'datapackage.json')))
pudl_engine = pudl.output.export.connect_db(pudl_settings=pudl_settings,
                                                testing=testing)
pkg.save(storage='sql', engine=pudl_engine)
```
I'm getting `CastError: There are 18 cast errors (see exception.errors)`.

Do you know if saving to sqlite with partitioned tables is supposed to be working? It may be something on my end but when I regenerated the one table without a partition, it worked fine.

**Comment by cmgosnell** [comment]
one more question for you @roll! when exporting to sqlite, I'm not getting any of the foreign keys being exported. It [looks like](https://github.com/frictionlessdata/pilot-catalyst#exporting-data-sql) they are being pulled over in postgres, but I'm having some issues getting the export to postgres working (it's failing on some fk relationships) and we were wanting to use sqlite.

**Comment by roll** [comment]
@cmgosnell 
Can you please post contents of `exception.errors`? 

Please try `pip install --upgrade tableschema-sql==1.0.1` for SQLite FK support

**Comment by cmgosnell** [comment]
@roll, here are the `exception.errors` from exporting a partitioned package to sqlite:
```
[CastError('Field "id" can\'t cast value "id" for type "integer" with format "default"'),
 CastError('Field "state" has constraint "enum" which is not satisfied for value "state"'),
 CastError('Field "plant_id_eia" can\'t cast value "plant_id_eia" for type "integer" with format "default"'),
 CastError('Field "operating_datetime_utc" can\'t cast value "operating_datetime_utc" for type "datetime" with format "default"'),
 CastError('Field "operating_time_hours" can\'t cast value "operating_time_hours" for type "number" with format "default"'),
 CastError('Field "gross_load_mw" can\'t cast value "gross_load_mw" for type "number" with format "default"'),
 CastError('Field "steam_load_1000_lbs" can\'t cast value "steam_load_1000_lbs" for type "number" with format "default"'),
 CastError('Field "so2_mass_lbs" can\'t cast value "so2_mass_lbs" for type "number" with format "default"'),
 CastError('Field "so2_mass_measurement_code" has constraint "enum" which is not satisfied for value "so2_mass_measurement_code"'),
 CastError('Field "nox_rate_lbs_mmbtu" can\'t cast value "nox_rate_lbs_mmbtu" for type "number" with format "default"'),
 CastError('Field "nox_rate_measurement_code" has constraint "enum" which is not satisfied for value "nox_rate_measurement_code"'),
 CastError('Field "nox_mass_lbs" can\'t cast value "nox_mass_lbs" for type "number" with format "default"'),
 CastError('Field "nox_mass_measurement_code" has constraint "enum" which is not satisfied for value "nox_mass_measurement_code"'),
 CastError('Field "co2_mass_tons" can\'t cast value "co2_mass_tons" for type "number" with format "default"'),
 CastError('Field "co2_mass_measurement_code" has constraint "enum" which is not satisfied for value "co2_mass_measurement_code"'),
 CastError('Field "heat_content_mmbtu" can\'t cast value "heat_content_mmbtu" for type "number" with format "default"'),
 CastError('Field "facility_id" can\'t cast value "facility_id" for type "integer" with format "default"'),
 CastError('Field "unit_id_epa" can\'t cast value "unit_id_epa" for type "integer" with format "default"')]
```

These are all of the columns from the `hourly_emmissions_epacems` table:
`id,state,plant_id_eia,unitid,operating_datetime_utc,operating_time_hours,gross_load_mw,steam_load_1000_lbs,so2_mass_lbs,so2_mass_measurement_code,nox_rate_lbs_mmbtu,nox_rate_measurement_code,nox_mass_lbs,nox_mass_measurement_code,co2_mass_tons,co2_mass_measurement_code,heat_content_mmbtu,facility_id,unit_id_epa`

Also, I updated to that `tableschema-sql` version and know I have fk's! yay,thanks.

**Comment by zaneselvans** [comment]
My guess here is that the header row in each of the CSV files that is not the first CSV file being loaded into the EPA CEMS table is being mistakenly interpreted as a record that should be pulled into the database, and it's failing because for many of them, the string which is in the header row does not satisfy the data type requirements for the column it's in.

**Comment by cmgosnell** [comment]
That sounds right because the only column that is not getting an error is `unitid` which is the only column of strings.

**Comment by zaneselvans** [comment]
It seems right to me that each of the CSV files that make up the partitions ought to have a header row though, so that they remain human readable and could be loaded against the same schema without additional information, or strictly requiring the exact same number and ordering of fields. I see a reference in the [Data Resource](http://frictionlessdata.io/specs/data-resource/) specification:

> NOTE: all files in the array MUST be similar in terms of structure, format etc. Implementors MUST be able to concatenate together the files in the simplest way and treat the result as one large file. For tabular data there is the issue of header rows. See the Tabular Data Package spec for more on this.

But I'm not seeing any "more" about the details of data in multiple files or the issue of non-data header rows in the [Tabular Data Package](http://frictionlessdata.io/specs/tabular-data-package/) or [Tabular Data Resource](http://frictionlessdata.io/specs/tabular-data-resource/) specifications.

**Comment by roll** [comment]
I don't see either. 

As an implementor, I read this sentence `Implementors MUST be able to concatenate together the files in the simplest way and treat the result as one large file. ` clearly as `files must be mechanically split into chunks` (meaning that onle the first file has a headers row). So in the libraries, I merge it at the byte stream level (much earlier than any headers analysis happens).

**Comment by roll** [comment]
@cmgosnell @zaneselvans 
Also, regarding bytes/hash, I want to reraise my question. What is the downside of providing bytes/hash of the raw uncompressed, unpartitioned file? As I see it's enough to verify the file is not compromised.

**Comment by roll** [comment]
> @cmgosnell @zaneselvans
> Also, regarding bytes/hash, I want to reraise my question. What is the downside of providing bytes/hash of the raw uncompressed, unpartitioned file? As I see it's enough to verify the file is not compromised.

@cmgosnell @zaneselvans 
Could you please take a look. I'm starting to work on the point we have discussed (compression/goodtables-fk/etc). So it will be great to clarify this one

**Comment by zaneselvans** [comment]
A couple of things come to mind:
* It would mean that you need to decompress and recompress all of the data in order to check the hash or the filesize. In the case of our CEMS data that'll be about 100GB passing through memory.
* The hash value will depend on the order in which the files are concatenated -- is the order of concatenation guaranteed to be identical every time?
* To generalize that concern: it feels like the process for calculating the size or hash of a single file is extremely well defined and repeatable, while I'm not as clear on whether the process that would be used to do it on a concatenation of N files would always give exactly the same result. Could people do it slightly differently and end up with different results, yielding false validation errors?
* Localizing the validation failure to a particular file would allow the user to know which subset of the data is potentially damaged, rather than simply invalidating the entire resource.

More generally -- as with a lack of header rows in partitioned CSVs -- if the partitioned resource has to be completely re-assembled in order to do anything with it, it seems like a lot of the potential utility of partitioning is lost. What were y'all seeing as the value of allowing large files to be broken up?  I was hoping/imagining it would work in a way similar to how parquet datasets, where the partitions allow more efficient querying against a subset of the overall dataset... I.e. if we partitioned the CEMS data by year and state, then someone who only needs (say) the 10 western US states that make up the western electricity grid, from 2010 to the present, they can read that portion of the data, without having to deal with everything else.

Maybe we should really just be using parquet files for the larger data, but then the connection between the metadata and the data itself becomes more opaque.  Or maybe having a single 10GB compressed CSV would work okay? It just seems unwieldy.

**Comment by roll** [comment]
> More generally -- as with a lack of header rows in partitioned CSVs -- if the partitioned resource has to be completely re-assembled in order to do anything with it, it seems like a lot of the potential utility of partitioning is lost.
Yes. That's actually what I think too. At the moment, FD software considers resource multiparting as a physical layer operation like compression, but in your case, it is more logical level operation because headers are going to be preserved, partitioning is going to happen by a logical key (year), etc.

So here's a question then: why don't publish partitions as separate resources with a shared schema? 
In this case, users will be able to access it separately (as e.g. partitioned SQL table) but also we can implement API to work with resources as a group like `package.get_resource_group('data_year_*').read()` which will be easier to implement rather than to deal with partitioned resources. Eliminating partitioning within one resource will solve a lot of problems like lack of support in goodtables, in non-Python libs, uncertainty on the specs level about bytes/hash, etc

PS.
Regarding `you need to decompress and recompress all of the data`, the ultimate goal of the FD tools is memory efficiency and minimization of IO operations. So if there are a multipart resource and a hash for it:
- (1) we will start file reading and decompressing (local or remote) from the first part
- (2) will be processing the byte stream **row by row**, including
  - incremental hash calculation
  - type casting
  - all the checks
  - etc
- (3) will concatenate the <N>th part at byte stream layer repeating (2)
- (4) will finish when all the parts are processed
So, actually, if there is a need to run goodtables on some multipart resource the original file's hash check will not add any overhead.

**Comment by roll** [comment]
@cmgosnell 
@zaneselvans 
So I've analyzed the approach we discussed yesterday and I don't see any major blockers:

1. The plan is to support descriptors like this:

```yml
resources:
  - name: cars-2017
    group: cars
    schema: 'path' or 'object' # use a path to have a shared schema
  - name: cars-2018
    group: cars
    schema: 'path' or 'object' # use a path to have a shared schema
```

2. In the exporting to a database code:

```python
package = Package('datapackage.json')
package.save(storage='sql', engine=engine, merge_groups=True)
```

3. In the database:

```sql
/d
cars # combining the data from both resources
```

4. Extending this approach later like:

```python
# reading in python
resource.read(keyed=True, merge_groups=True)
# partitioning for from a database import
# ...
# etc
```

PS.
Naming for the fields/options is WIP.

**Comment by roll** [comment]
If it makes sense I will be implementing it tomorrow which shouldn't take long. So it's expected to be ready on the US Wednesday to unblock this issue

**Comment by zaneselvans** [comment]
I think that generally makes sense to me.  In the case of a shared schema, it would need to be its own separate file stored locally within the datapackage directory somewhere, right (or a URL in theory).

And in the `resource.read()` the `resource` object is just... one of the several resources that are part of the group? But by telling it to use merge the groups, it knows to go and get everything else that's part of the group? That seems like it requires the `resource` to be aware of the `datapackage` context that it exists within.  Is that right?

**Comment by cmgosnell** [comment]
this also generally sounds good to me. Just to ensure my understanding, if I were to run `package.save(storage='sql', engine=engine, merge_groups=False)` each of the resources would be exported as their own unique table, but with `merege_groups=True` it will take all of the resources with the same group name?

On the path to the schema, I have the same question as Zane. Do I need to store the resource descriptor separately/alone or could I populate one of the resource descriptors of a group in a metadata file and refer all of the others to that one?

**Comment by roll** [comment]
@cmgosnell 
@zaneselvans 
Please take a look and test if it works to solve this issue:

```bash
pip install --upgrade git+https://github.com/frictionlessdata/datapackage-py.git@groups-of-resources
```

Readme - https://github.com/frictionlessdata/datapackage-py/tree/groups-of-resources#group
Examples - https://github.com/frictionlessdata/datapackage-py/pull/243/files (see examples/tests)

**Comment by roll** [comment]
Please let me know what you think and whether it resolved the problem. If it's good to go I will release it tomorrow.

**Comment by roll** [comment]
Hi @cmgosnell,

I'm not sure about the first error. It seems like errors which happen when you create, for example` a `dapackage.py` file in the working directory so the script trying to import from this module instead of the real one. Could you please check?

Regarding the second one, I need a little bit more information. There is, for example, a datapackage witch combined compressed and plain resources:

- https://github.com/frictionlessdata/datapackage-py/tree/master/data/datapackage-groups
- https://raw.githubusercontent.com/frictionlessdata/datapackage-py/groups-of-resources/data/datapackage-groups/datapackage.json

I released as `datapackage@1.8.0` for simplicity so starting from a fresh virtual environment works for me for the datapackage above:

```bash
virtualenv .python -ppython3.7
source .python/bin/activate
pip install datapackage==1.8.0
pip install tableschema_sql
```

```python
import sqlalchemy
from datapackage import Package

# Save to SQL
engine = sqlalchemy.create_engine('sqlite://')
package = Package('https://raw.githubusercontent.com/frictionlessdata/datapackage-py/master/data/datapackage-groups/datapackage.json')
package.save(storage='sql', engine=engine, merge_groups=True)

# Read from SQL
print(list(engine.execute("select name from sqlite_master where type='table'")))
print(list(engine.execute("select * from cars")))
```

Could you please try and also post what you're running if it still fails.

**Comment by roll** [comment]
> it is assuming the csv files need to be in the same directory as the datapackage.json file

Actually, no. The idea that `resource.group` works on top all the datapackage functionality so you can define any data package you want (having different resources local/remote/compressed/etc) and just then like them with the `resource.group` field.

**Comment by roll** [comment]
Just of the top of my head, an example with inline data:

```python
import sqlalchemy
from datapackage import Package

# Save to SQL
package = Package({
    'resources': [
        {
            'group': 'cars',
            'name': 'cars-2016',
            'data': [
                ['name', 'value'],
                ['tesla', '2016'],
            ],
        },
        {
            'group': 'cars',
            'name': 'cars-2017',
            'data': [
                ['name', 'value'],
                ['tesla', '2017'],
            ],
        },
    ]
})
package.infer()
engine = sqlalchemy.create_engine('sqlite://')
package.save(storage='sql', engine=engine, merge_groups=True)

# Read from SQL
print(list(engine.execute("select name from sqlite_master where type='table'")))
# [('cars',)]
print(list(engine.execute("select * from cars")))
# [('tesla', 2016), ('tesla', 2017)]
```

**Comment by roll** [comment]
@cmgosnell 
It's possible with the `autoincrement` parameter - https://github.com/frictionlessdata/tableschema-sql-py#storageengine-dbschemanone-prefix-reflect_onlynone-autoincrementfalse:

```python
import sqlalchemy
from datapackage import Package

# Save to SQL
package = Package({
    'resources': [
        {
            'group': 'cars',
            'name': 'cars-2016',
            'data': [
                ['name', 'value'],
                ['tesla', '2016'],
            ],
        },
        {
            'group': 'cars',
            'name': 'cars-2017',
            'data': [
                ['name', 'value'],
                ['tesla', '2017'],
            ],
        },
    ]
})
package.infer()
engine = sqlalchemy.create_engine('postgresql://roll:roll@localhost:5432/datapackage')
package.save(storage='sql', engine=engine, merge_groups=True, autoincrement='id')

# Read from SQL
print(list(engine.execute("select table_name from information_schema.tables where table_schema='public'")))
# [('cars',)]
print(list(engine.execute("select * from cars")))
# [(1, 'tesla', 2016), (2, 'tesla', 2017)]
```

There are two problems:
- for now, it will add this field for all the data package table. We can solve it adding support for a mapping to specify resources with autoincrements `package.save(storage='sql', engine=engine, merge_groups=True, autoincrement={'resource1': 'id'})` - which is easy
- for now, it only works for PostgreSQL. SQLite has a lot of caveats and autoincrementing fields is one of them. I wasn't able to find a simple solution to how to handle it with `sqlalchemy`. SQLite doesn't support `insert.returning` etc. - but probably it's also doable but feels harder

**Comment by zaneselvans** [comment]
Hmm. The documentation on [SQLite autoincrementors in SQLAlchemy](https://docs.sqlalchemy.org/en/13/dialects/sqlite.html#sqlite-auto-incrementing-behavior) makes it sound like it should be possible to just add a `primary_key` field of type `INTEGER PRIMARY KEY` to the table definition if it's requested by the `autoincrement` dictionary you've suggested and have it do the right thing. Am I not understanding it right?

**Comment by roll** [comment]
@zaneselvans 
Sure it should work probably it just has different API on the sqlalchemy level and doesn't support `returning` which we rely heavily on as I see - https://github.com/frictionlessdata/tableschema-sql-py/blob/master/tableschema_sql/writer.py 

But, ok, I think I will make it work during this week

**Comment by zaneselvans** [comment]
Thank you @roll! I hope this all hasn't been too much more work than we all bargained for.  I think this is the last issue keeping us from deprecating the postgres database, and merging our development branches into `master`.  Looking forward to getting our users some easily installable data.

Do you work on the datahub.io side of things at all? How has development on the [data installation CLI](https://github.com/datopian/data-cli) progressed?  It looks like it hasn't seen much attention since last fall.

**Comment by roll** [comment]
@cmgosnell
@zaneselvans 
Here we go with `tableschema-sql@1.1.0`:

```python
import sqlalchemy
from datapackage import Package

# Package

package = Package({
    'resources': [
        {
            'name': 'cities',
            'data': [
                ['name', 'value'],
                ['london', '2016'],
                ['london', '2017'],
            ],
        },
        {
            'group': 'cars',
            'name': 'cars-2016',
            'data': [
                ['name', 'value'],
                ['tesla', '2016'],
            ],
        },
        {
            'group': 'cars',
            'name': 'cars-2017',
            'data': [
                ['name', 'value'],
                ['tesla', '2017'],
            ],
        },
    ]
})
package.infer()

# Database

engine = sqlalchemy.create_engine('sqlite://')
package.save(storage='sql', engine=engine, merge_groups=True, autoincrement={'cars': 'id'})
print(list(engine.execute("select name from sqlite_master where type='table'")))
# [('cars',)]
print(list(engine.execute("select * from cities")))
# [('london', 2016), ('london', 2017)]
print(list(engine.execute("select * from cars")))
# [(1, 'tesla', 2016), (2, 'tesla', 2017)]
```

PS.
Regarding the second question, we (Open Knowledge) have no involvement in datahub.io development, actually. So I'm not sure.

**Comment by zaneselvans** [comment]
So, without the autoincrementing ID fields in the data tables themselves, some of our tables have valid duplicate rows, which by default are prohibited by goodtables. We can use the `skip_checks` argument in our own validation of the data package to bypass this test when it's appropriate, but unless that information is stored in the tableschema (or elsewhere in the tabular data resource metadata) when a user attempts to validate the package, it will fail.  Is there some way to specify this kind of constraint (or lack of constraint) in a TableSchema? It seems similar to the per-field constraints that can be specified, but operating at the level of the whole table. It doesn't seem like this is contained in the current specification.

**Comment by roll** [comment]
@zaneselvans 
Yes. There is no way to specify it within the spec. So probably only adding a note to the description of the data package. 

In http://goodtables.io/ it's possible to disable checks using `config.yml` but it doesn't cover all the use cases, of course (when a user uses CLI, etc).

**Comment by cmgosnell** [comment]
Hey @roll! Thanks for all of these updates! I incorporated the autoincrement into all of  our tables (except for one) that were using indexes as `id` columns and it worked great. So that is lovely.

On the duplicate records and using `skip_checks`, right now I have it all working just using the index like I was before. If we were to do this, it would be important to be able to be able to use `skip_checks` for specific tables while validating the whole package. I'm not sure that this is critical right now, but we'll have to think about if there are any other implications of the current set up.

**Comment by cmgosnell** [comment]
I'm going to close this issue because the sqlite export seems to be doing it's thing. But, I'm going to make a new issue to discuss `skip_checks`...
